From a68d66ca342d37650b7c028334f6a86a588f0d53 Mon Sep 17 00:00:00 2001
From: Mohit Sabharwal <mohit@cloudera.com>
Date: Thu, 26 May 2016 23:36:05 -0700
Subject: [PATCH 0611/1164] CDH-40070: HIVE-13836 : DbNotifications giving an
 error = Invalid state.  Transaction has already
 started Note that this is an optimisitic
 backport.

Change-Id: I5bdd6970d456dd495054f2435eddd0283750613f
---
 .../hcatalog/listener/DbNotificationListener.java  |   10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/hcatalog/server-extensions/src/main/java/org/apache/hive/hcatalog/listener/DbNotificationListener.java b/hcatalog/server-extensions/src/main/java/org/apache/hive/hcatalog/listener/DbNotificationListener.java
index f2e70ed..9181d89 100644
--- a/hcatalog/server-extensions/src/main/java/org/apache/hive/hcatalog/listener/DbNotificationListener.java
+++ b/hcatalog/server-extensions/src/main/java/org/apache/hive/hcatalog/listener/DbNotificationListener.java
@@ -63,6 +63,8 @@
   private static final Log LOG = LogFactory.getLog(DbNotificationListener.class.getName());
   private static CleanerThread cleaner = null;
 
+  private static final Object NOTIFICATION_TBL_LOCK = new Object();
+
   // This is the same object as super.conf, but it's convenient to keep a copy of it as a
   // HiveConf rather than a Configuration.
   private HiveConf hiveConf;
@@ -258,7 +260,9 @@ private int now() {
 
   private void enqueue(NotificationEvent event) {
     if (rs != null) {
-      rs.addNotificationEvent(event);
+      synchronized(NOTIFICATION_TBL_LOCK) {
+        rs.addNotificationEvent(event);
+      }
     } else {
       LOG.warn("Dropping event " + event + " since notification is not running.");
     }
@@ -280,7 +284,9 @@ private void enqueue(NotificationEvent event) {
     @Override
     public void run() {
       while (true) {
-        rs.cleanNotificationEvents(ttl);
+        synchronized(NOTIFICATION_TBL_LOCK) {
+          rs.cleanNotificationEvents(ttl);
+        }
         try {
           Thread.sleep(60000);
         } catch (InterruptedException e) {
-- 
1.7.9.5

